engines/private/private.cpp

```diff
@@ -0,0 +1,1081 @@
+        cframe->free();
+        //delete frame;
+        delete cframe;
+    }
+
+    Graphics::Surface *screen = g_system->lockScreen();
+    if (_mode == 1) {
+        drawScreenFrame(screen);
+    }
+    Common::Rect window(_origin->x, _origin->y, _screenW - _origin->x, _screenH - _origin->y);
+    screen->copyRectToSurface(*surface, _origin->x, _origin->y, window);
+    g_system->unlockScreen();
+    //if (_image->getPalette() != nullptr)
+    //    g_system->getPaletteManager()->setPalette(_image->getPalette(), _image->getPaletteStartIndex(), _image->getPaletteColorCount());
+    g_system->updateScreen();
+
+}
+
+bool PrivateEngine::getRandomBool(uint p) {
+    uint r = _rnd->getRandomNumber(100);
+    return (r <= p);
+}
+
+Common::String *PrivateEngine::getPaperShuffleSound() {
+    uint r = _rnd->getRandomNumber(6);
+    char f[32];
+    snprintf(f, 32, "glsfx0%d.wav", kPaperShuffleSound[r]);
+    return (new Common::String(*_globalAudioPath + f));
+}
+
+Common::String *PrivateEngine::getTakeSound() {
+    if (isDemo())
+        return (new Common::String(*_globalAudioPath + "mvo007.wav"));
+
+    uint r = _rnd->getRandomNumber(4) + 1;
+    char f[32];
+    snprintf(f, 32, "took%d.wav", r);
+    return (new Common::String(*_globalAudioPath + f));
+}
+
+Common::String *PrivateEngine::getTakeLeaveSound() {
+    uint r = _rnd->getRandomNumber(1);
+    if (r == 0) {
+        return (new Common::String(*_globalAudioPath + "mvo001.wav"));
+    } else {
+        return (new Common::String(*_globalAudioPath + "mvo006.wav"));
+    }
+}
+
+Common::String *PrivateEngine::getLeaveSound() {
+    if (isDemo())
+        return (new Common::String(*_globalAudioPath + "mvo008.wav"));
+
+    uint r = _rnd->getRandomNumber(4) + 1;
+
+    char f[32];
+    snprintf(f, 32, "left%d.wav", r);
+    return (new Common::String(*_globalAudioPath + f));
+}
+
+char *PrivateEngine::getRandomPhoneClip(char *clip, int i, int j) {
+    uint r = i + _rnd->getRandomNumber(j - i);
+
+    char *f = (char *)malloc((strlen(clip)+3)*sizeof(char));
+    snprintf(f, 32, "%s%02d", clip, r);
+    return f;
+}
+
+// Timers
+void timerCallback(void *refCon) {
+    g_private->removeTimer();
+    g_private->_nextSetting = (Common::String *)refCon;
+}
+
+bool PrivateEngine::installTimer(uint32 delay, Common::String *ns) {
+    return g_system->getTimerManager()->installTimerProc(&timerCallback, delay, (void *)ns, "timerCallback");
+}
+
+void PrivateEngine::removeTimer() {
+    g_system->getTimerManager()->removeTimerProc(&timerCallback);
+}
+
+// Diary
+
+void PrivateEngine::loadLocations(Common::Rect *rect) {
+    uint32 i = 0;
+    int16 offset = 44;
+    for (NameList::iterator it = locationList.begin(); it != locationList.end(); ++it) {
+        Private::Symbol *sym = locations.getVal(*it);
+        i++;
+        if (sym->u.val) {
+            offset = offset + 22;
+            char *f = (char *)malloc(32*sizeof(char));
+            snprintf(f, 32, "dryloc%d.bmp", i);
```

Comment:
Use `Common::String::format()` and in other similar places. You're not freeing `f` anywhere anyway.

Reply:
I think this is fixed. The only use of `snprintf` remaining is in `getRandomPhoneClip`.